File: packages/hpack-html-py/hpack_html/varint.py
"""Protobuf-style variable-length integer encoding/decoding."""

MAX_VARINT_BYTES = 7


def encode_varint(value: int) -> bytes:
    """Encode an unsigned integer as a VarInt byte sequence."""
    if value < 0:
        raise ValueError(f"VarInt value must be non-negative, got {value}")
    if value == 0:
        return b"\x00"

    result = bytearray()
    remaining = value
    while remaining > 0:
        byte = remaining & 0x7F
        remaining >>= 7
        if remaining > 0:
            byte |= 0x80
        result.append(byte)
    return bytes(result)


def decode_varint(data: bytes | bytearray | memoryview, offset: int) -> tuple[int, int]:
    """Decode a VarInt from data at offset. Returns (value, bytes_read)."""
    if offset < 0 or offset >= len(data):
        raise ValueError(f"VarInt offset out of bounds: {offset} (data length: {len(data)})")

    value = 0
    shift = 0
    bytes_read = 0

    while True:
        pos = offset + bytes_read
        if pos >= len(data):
            raise ValueError(f"VarInt truncated at offset {offset} (read {bytes_read} bytes)")

        byte = data[pos]
        bytes_read += 1
        value |= (byte & 0x7F) << shift
        shift += 7

        if byte & 0x80 == 0:
            break
        if bytes_read >= MAX_VARINT_BYTES:
            raise ValueError(f"VarInt exceeds maximum of {MAX_VARINT_BYTES} bytes")

    return value, bytes_read

File: packages/hpack-html-go/varint.go
package hpack

import "fmt"

const maxVarintBytes = 7

// DecodeVarInt decodes a protobuf-style VarInt from data at offset.
// Returns (value, bytesRead, error).
func DecodeVarInt(data []byte, offset int) (uint64, int, error) {
	if offset < 0 || offset >= len(data) {
		return 0, 0, fmt.Errorf("varint offset out of bounds: %d (data length: %d)", offset, len(data))
	}

	var value uint64
	var shift uint
	bytesRead := 0

	for {
		pos := offset + bytesRead
		if pos >= len(data) {
			return 0, 0, fmt.Errorf("varint truncated at offset %d (read %d bytes)", offset, bytesRead)
		}

		b := data[pos]
		bytesRead++

		value |= uint64(b&0x7F) << shift
		shift += 7

		if b&0x80 == 0 {
			break
		}
		if bytesRead >= maxVarintBytes {
			return 0, 0, fmt.Errorf("varint exceeds maximum of %d bytes", maxVarintBytes)
		}
	}

	return value, bytesRead, nil
}

// EncodeVarInt encodes a uint64 as a VarInt byte sequence.
func EncodeVarInt(value uint64) []byte {
	if value == 0 {
		return []byte{0}
	}
	var buf []byte
	remaining := value
	for remaining > 0 {
		b := byte(remaining & 0x7F)
		remaining >>= 7
		if remaining > 0 {
			b |= 0x80
		}
		buf = append(buf, b)
	}
	return buf
}
